fix(lifecycle): platform check for default install scripts

reyronald opened the pull request:
Would be nice if @caleblloyd could jump in and test this again in his environment, just in case.

Naturally, I'm not familiar with the rest of the code base and npm's internals, so I can't be certain if there are no other side-effects that should be considered. I'm all ears. 

Fixes #49
Related to #45, #46

reyronald committed:
fix(lifecycle): platform check for default install scripts

This is a continuation of #45, a side-effect of the provided fix for it
in #46 and was introduced in `cipm@1.6.2`.

For the case where a default `install` script is used and a
`binding.gyp` file is present in  the package root, `npm ci` will fail
with packages that target a different platform that the one currently
running.

Fixes #49

zkat commented on index.js:
```diff
@@ -317,17 +318,21 @@ class Installer {
+          .then(pkg)
       })
-      .then(pkg)
+      .catch(() => 'ignore')
```
If you look at the `buildTree()` method, you'll notice a `if (!checkDepEnv(...)) {...}`. What that does is filter out the dependency altogether while building the tree we're going to extract. If you add the necessary checks to `checkDepEnv()`, you should be good.

Interesting to note that `checkPlatform()` is actually synchronous, even though it takes a callback: https://github.com/npm/npm-install-checks/blob/master/index.js#L27-L54 so I'd just rather you write a wrapper that treats it as sync so you don't have to refactor all the `checkDepEnv` calls into async calls.

zkat commented on test/specs/index.js:
```diff
@@ -718,11 +727,14 @@ test('adds install script when binding.gyp is present', t => {
     }, 'dep b binary is not built')
+    t.throws(() => {
+      fs.statSync(path.join(prefix, 'node_modules', 'c', 'build', 'Release', 'hello' + binarySuffix))
+    }, 'dep c binary is not built')
```
hooray for tests! v. v. happy about this 💯

reyronald committed:
adress pull request suggested changes

- Sync wrapper for `checkPlatform`
- Move `checkPlatform` check from `updateInstallScript` to `checkDepEnv`

reyronald commented:
@zkat I had to add `pkg` parameter to `checkDepEnv` to be able to perform the platform check inside it. It is optional though because the only place where it is readily available before the call is in `buildTree`. In all the other places (`extractTree` and `updateJson`) the `package.json` is not read yet at that point and I didn't wanna go ahead and do such a drastic change. So it feels kind of ugly to me this way, did you consider this?

Perhaps it would be best to call `checkPlatformSync` directly in `buildTree` right next to`checkDepEnv`, instead of having it inside it and having to do that null/undefined check and bypass it like it is now ? Honestly I don't mind much either way so if you think it's fine then 👍 , unless you have another suggestion, just wanted to mention it just in case.

P.S.: Should we include that `checkPlatformSync` function in the `npm-install-checks` package itself ?

reyronald committed:
Configure vscode's debugging env

evocateur commented on test/specs/index.js:
```diff
@@ -657,7 +658,7 @@ test('skips lifecycle scripts with ignoreScripts is set', t => {
 })
 
-test('adds install script when binding.gyp is present', t => {
+only('adds install script when binding.gyp is present', t => {
```
I would imagine this `only()` needs to be a `test()` again before this PR can be merged...

zkat committed:
Delete launch.json

zkat committed:
Delete tasks.json
